src/snowflake/telemetry/_internal/exporter/otlp/proto/log_exporter/__init__.py

```diff
@@ -0,0 +1,170 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import abc
+import enum
+import logging
+import logging.config
+import threading
+import typing
+import opentelemetry.sdk.util.instrumentation as otel_instrumentation
+
+from opentelemetry.sdk import resources
+from opentelemetry.sdk._logs import export
+from opentelemetry.sdk import _logs
+from opentelemetry.util import types
+from snowflake.telemetry._internal.encoder.otlp.proto.common.log_encoder import (
+    serialize_logs_data
+)
+
+LOGGER_NAME_TEMP_ATTRIBUTE = "__snow.logging.temp.logger_name"
+FILEPATH_ATTRIBUTE = "code.filepath"
+FUNCTION_NAME_ATTRIBUTE = "code.function"
+
+
+class LogWriterResult(enum.Enum):
+    SUCCESS = 0
+    FAILURE = 1
+
+
+# LogWriter abstract class with one abstract method that can be overwritten:
+class LogWriter(abc.ABC):
+
+    @abc.abstractmethod
+    def write_logs(self, serialized_logs: bytes) -> LogWriterResult:
+        """quick"""
+
+
+class _ProtoLogExporter(export.LogExporter):
+
+    def __init__(self, log_writer: LogWriter):
+        self.log_writer = log_writer
+
+    def export(self, batch: typing.Sequence[_logs.LogData]):
+        result = self.log_writer.write_logs(serialize_logs_data(batch))
+        if result == LogWriterResult.FAILURE:
+            return export.LogExportResult.FAILURE
+        return export.LogExportResult.SUCCESS
+
+
+    def shutdown(self):
+        pass
+
+
+class SnowflakeLoggingHandler(_logs.LoggingHandler):
+    """A subclass of OpenTelemetry's LoggingHandler that preserves attributes discarded by the original implementation."""
+
+    def __init__(
+            self,
+            log_writer: LogWriter,
+        ):
+        provider = _SnowflakeTelemetryLogEmitterProvider()
+        _logs.set_log_emitter_provider(provider)
+        exporter = _ProtoLogExporter(log_writer)
+        provider.add_log_processor(export.SimpleLogProcessor(exporter))
+        super().__init__()
+
+    @staticmethod
+    def _getSnowflakeLogLevelName(pyLevelName):
+        """Adapted from the getSnowflakeLogLevelName in XP's logger.py"""
+        level = pyLevelName.upper()
+        if level == "WARNING":
+            return "WARN"
+        elif level == "CRITICAL":
+            return "FATAL"
+        elif level == "NOTSET":
+            return "TRACE"
+        else:
+            return level
+
+    def _get_attributes(self, record: logging.LogRecord) -> types.Attributes:
+        attributes = super(SnowflakeLoggingHandler, self)._get_attributes(record)
+
+        # Adding attributes that were discarded by the base class
```

Comment:
Consider to also add record.lineno as `code.lineno` (see https://opentelemetry.io/docs/specs/semconv/attributes-registry/code/). If you have questions about this please talk directly to me.

Reply:
Created a Jira to track this here: https://snowflakecomputing.atlassian.net/browse/SNOW-1215215